source/common/event/real_time_system.cc

```diff
@@ -0,0 +1,63 @@
+#include "common/event/real_time_system.h"
+
+#include <chrono>
+
+#include "common/common/assert.h"
+#include "common/event/event_impl_base.h"
+
+#include "event2/event.h"
+
+namespace Envoy {
+namespace Event {
+namespace {
+
+/**
+ * libevent implementation of Timer.
+ */
+class TimerImpl : public Timer, ImplBase {
+public:
+  TimerImpl(Libevent::BasePtr& libevent, TimerCb cb) : cb_(cb) {
+    ASSERT(cb_);
+    evtimer_assign(
+        &raw_event_, libevent.get(),
+        [](evutil_socket_t, short, void* arg) -> void { static_cast<TimerImpl*>(arg)->cb_(); },
+        this);
+  }
+
+  // Timer
+  void disableTimer() override { event_del(&raw_event_); }
+  void enableTimer(const std::chrono::milliseconds& d) override {
+    if (d.count() == 0) {
+      event_active(&raw_event_, EV_TIMEOUT, 0);
+    } else {
+      std::chrono::microseconds us = std::chrono::duration_cast<std::chrono::microseconds>(d);
+      timeval tv;
+      tv.tv_sec = us.count() / 1000000;
```

Comment:
The conversion factors here could be more readable by using duration cast:

```
auto secs = std::chrono::duration_cast<std::chrono::seconds>(d);
auto usecs = std::chrono::duration_cast<std::chrono::microseconds>(d - secs);
tv.tv_secs = secs.count();
tv.tv_usecs = usecs.count();
```

Reply:
Agreed, but I have strictly moved this code verbatim from timer_impl.h and would prefer not to rewrite it in this PR.

Reply:
Actually I decided to restore timer_impl.h and timer_impl.cc which reveals the 'verbatim' note above to be a lie; there is a slight change which is more visible now (the timers are constructed with a libevent base rather than a dispatcher).

I did this in anticipation of delegating to real-time TimerImpl when building simulated-time TimerImpl in the next PR, but it also makes this diff cleaner :). In simulated time, the real-time timer delegate won't be enabled until the simulated time has advanced enough for it to fire, but I still want it to fire via the libevent dispatcher mechanism.